rotkehlchen/exchanges/bitcoinde.py

```diff
@@ -0,0 +1,249 @@
+        timestamp = deserialize_timestamp_from_date(
+            raw_trade['trade_marked_as_paid_at'],
+            'iso8601',
+            'bitcoinde',
+        )
+
+    trade_type = deserialize_trade_type(raw_trade['type'])
+    tx_amount = AssetAmount(FVal(raw_trade['amount_currency_to_trade']))
+    native_amount = FVal(raw_trade['volume_currency_to_pay'])
+    tx_asset, native_asset = bitcoinde_pair_to_world(raw_trade['trading_pair'])
+    pair = TradePair(f'{tx_asset.identifier}_{native_asset.identifier}')
+    amount = tx_amount
+    rate = Price(native_amount / tx_amount)
+    fee_amount = deserialize_fee(raw_trade['fee_currency_to_pay'])
+    fee_asset = Asset('EUR')
+
+    return Trade(
+        timestamp=timestamp,
+        location=Location.BITCOINDE,
+        pair=pair,
+        trade_type=trade_type,
+        amount=amount,
+        rate=rate,
+        fee=fee_amount,
+        fee_currency=fee_asset,
+        link=str(raw_trade['trade_id']),
+    )
+
+
+class Bitcoinde(ExchangeInterface):
+    def __init__(
+            self,
+            api_key: ApiKey,
+            secret: ApiSecret,
+            database: 'DBHandler',
+            msg_aggregator: MessagesAggregator,
+    ):
+        super().__init__('bitcoinde', api_key, secret, database)
+        self.uri = 'https://api.bitcoin.de'
+        self.session.headers.update({'x-api-key': api_key})
+        self.msg_aggregator = msg_aggregator
+
+    def _generate_signature(self, request_type: str, url: str, nonce: str) -> str:
+        md5_empty = 'd41d8cd98f00b204e9800998ecf8427e'  # this is md5('')
+        signed_data = '#'.join([request_type, url, self.api_key, nonce, md5_empty]).encode()
+        signature = hmac.new(
+            self.secret,
+            signed_data,
+            hashlib.sha256,
+        ).hexdigest()
+        self.session.headers.update({
+            'x-api-signature': signature,
+        })
+        return signature
+
+    def _api_query(
+            self,
+            verb: str,
+            path: str,
+            options: Optional[Dict] = None,
+    ) -> Dict:
+        """
+        Queries Bitcoin.de with the given verb for the given path and options
+        """
+        assert verb in ('get', 'post', 'push'), (
+            'Given verb {} is not a valid HTTP verb'.format(verb)
+        )
+
+        request_path_no_args = '/v4/' + path
+
+        data = ''
+        if not options:
+            request_path = request_path_no_args
+        else:
+            request_path = request_path_no_args + '?' + urlencode(options)
+
+        nonce = str(int(time.time() * 1000))
+        request_url = self.uri + request_path
+
+        self._generate_signature(
+            request_type=verb.upper(),
+            url=request_url,
+            nonce=nonce,
+        )
+
+        self.session.headers.update({
+            'x-api-nonce': nonce,
+        })
+        if data != '':
+            self.session.headers.update({
```

Comment:
There may be a bug hidden here. We had it in some other exchange and @vnavascues spotted it.

Since `self.session.headers` keeps state in between calls, a call may have headers from a previous api call unless it's cleaned.

Depending on the headers and the way the Remote server behaves this may end up in an error.

For example here imagine the following scenario:

1. A call with data of length 100 happens succesfully.
2. Then you do a call without data, so we never get in here, but the headers of the previous session are still active. Bitcoin.de server may process this incorrectly.

I guess the safest approach is to determine which headers should stay session-wide, and populate them in the beginning (the exchange constructor). And then in each call append the call specific headers.

Note: This buggy behaviour may still be present in some of our exchange integrations. @vnavascues fixed it for one or two but he also wanted to check the others to make sure.

Reply:
Moved all headers to an argument to `get` as they are request-specific